### PR TITLE
Enhance to record more accurate e2e provision/deletion latency metric

### LINKED ISSUE #77030: Expose end to end latency volume operation metrics via CSI
**What would you like to be added**:
Latency of volume operations like provisioning a volume has been evaluated as following:
1. for in-tree implementation, latency is recorded ONLY for one-retry loop which could be not accurate, i.e., when the first retry actually fails while the following attempt succeeded.
2. via CSI, as its async nature, i.e. [https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/persistentvolume/pv_controller.go#L1404], the call for provisioning will return immediately and making the latency metrics recording at [https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/persistentvolume/pv_controller.go#L1348](url) not accurate.

This issue has been brought up in [https://github.com/kubernetes/kubernetes/issues/75782](url), creating a separate issue to track the effort for enhance it in the current controller after discussion with @msau42 as itself seems to be big enough.

Within this enhancement, looking to have two stages:
1. provide accurate volume provisioning/deletion latency metrics for CSI workflow
2. discuss with sig/storage to see whether it is worth enhancing existing in-tree implementation

**Why is this needed**:
To provide more accurate volume operation latency metrics for provisioning/deletion for a better user experience.

### LINKED ISSUE #77608: Total provisioning metrics are not correct
The total provisioning time metrics introduced here - https://github.com/kubernetes/kubernetes/pull/66389 are not acurate. This is because, they only measure single execution of provisioning(and deletion) and do not count retries. 

These should instead time, total time taken when first PVC was created and time when PVC was actually bound. 

/sig storage
cc @bertinatto @msau42

### PR DESCRIPTION
**Pull request by yuxiangqian**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Record accurate end to end latency for provisioning/deletion into metric name "volume_operation_total_time"
**Which issue(s) this PR fixes**:

Fixes #77030
Fixes #77608

**Special notes for your reviewer**:
unit tests created

**Does this PR introduce a user-facing change?**:
Metrics (end to end latency/error count) for provisioning/deletion have not been correctly recorded.
1. Time taken by external provisioner to create a volume has not been taken into consideration.
2. for in-tree workflow, the metric record the time spent on the first try of the operation instead of waiting for it to complete.
This PR is to enhance pv controller to record more accurate latency/error count metrics for users to consume.
```release-note
Introducing new semantic for metric "volume_operation_total_seconds" to be the end to end latency of volume provisioning/deletion. Existing metric "storage_operation_duration_seconds" will remain untouched however exposed to the following potential issues:
1. for volume's provisioned/deleted via external provisioner/deleter, "storage_operation_duration_seconds" will NOT wait for the external operation to be done before reporting latency metric (effectively close to 0). This will be fixed by using "volume_operation_total_seconds" instead
2. if there's a transient error happened during "provisioning/deletion", i.e., a volume is still in-use while a deleteVolume has been called, original "storage_operation_duration_seconds" will NOT wait until a volume has been finally deleted before reporting a not accurate latency metric. The newly implemented metric "volume_operation_total_seconds", however, wait util a provisioning/deletion operation has been fully executed.

Potential impacts:
If an SLO/alert has been defined based on "volume_operation_total_seconds", it might get violated because of the more accurate metric might be significantly larger than previously reported. The metric is defined to be a histogram and the new semantic could change the distribution.
```

### CODE REVIEW

File: pkg/controller/volume/persistentvolume/pv_controller.go [hunk 1]
```diff
@@ -323,13 +345,21 @@ func (ctrl *PersistentVolumeController) syncUnboundClaim(claim *v1.PersistentVol
...
+				// record count error for provision if exists, do not clean up cache
+				metrics.RecordMetric(claimKey, &ctrl.operationTimestamps, err)
 				return err
 			}
 			// OBSERVATION: claim is "Bound", pv is "Bound"
+			// if exists a timestamp entry in cache, record end to end provision latency and clean up cache
+			// End of the provision + binding operation lifecycle
+			// [Unit test 12-1, 12-2, 12-4]
+			metrics.RecordMetric(claimKey, &ctrl.operationTimestamps, nil)
+			ctrl.operationTimestamps.Delete(claimKey)
```
Comment (gnufied) [inline pkg/controller/volume/persistentvolume/pv_controller.go]: Do you think we could move `RecordMetric` as a function inside `operationTimestamps` - so as act of recording the metric also automatically deletes the key from cache? It will less error prone that way?

File: pkg/controller/volume/persistentvolume/pv_controller_base.go [hunk 2]
```diff
@@ -245,20 +251,26 @@ func (ctrl *PersistentVolumeController) updateClaim(claim *v1.PersistentVolumeCl
...
+// Unit test [5-5] [5-6] [5-7]
 // deleteClaim runs in worker thread and handles "claim deleted" event.
 func (ctrl *PersistentVolumeController) deleteClaim(claim *v1.PersistentVolumeClaim) {
 	_ = ctrl.claims.Delete(claim)
-	klog.V(4).Infof("claim %q deleted", claimToClaimKey(claim))
+	claimKey := claimToClaimKey(claim)
+	klog.V(4).Infof("claim %q deleted", claimKey)
+	// clean any possible unfinished provision start timestamp from cache
+	// Unit test [5-8] [5-9]
+	ctrl.operationTimestamps.Delete(claimKey)
```
Comment (gnufied) [inline pkg/controller/volume/persistentvolume/pv_controller_base.go]: same as before

File: pkg/controller/volume/persistentvolume/framework_test.go [hunk 3]
```diff
@@ -419,6 +419,26 @@ func claimWithAnnotation(name, value string, claims []*v1.PersistentVolumeClaim)
...
+}
+
+// volumesWithAnnotation saves given annotation into given volumes.
+// Meant to be used to compose volumes specified inline in a test.
+func volumesWithAnnotation(name, value string, volumes []*v1.PersistentVolume) []*v1.PersistentVolume {
+	for _, volume := range volumes {
+		volumeWithAnnotation(name, value, volume)
+	}
+	return volumes
+}
```
Comment (gnufied) [inline pkg/controller/volume/persistentvolume/framework_test.go]: Is any code uses these functions?

Comment (gnufied) [reply pkg/controller/volume/persistentvolume/framework_test.go]: ah nvm. I see it now.

File: pkg/controller/volume/persistentvolume/pv_controller.go (hunk 1 above)
Comment (yuxiangqian) [reply pkg/controller/volume/persistentvolume/pv_controller.go]: agreed leaving the life-cycle management of a cache item to the caller, i.e., in this case pv_controller/pv_controller_base, may causes cache entry leaking if caller forgot to call Delete. I propose to manage the deletion within RecordMetric, however not letting that function inside OperationStartTimeCache but still under metrics.go instead. To me, the cache should not be dealing with cache entry lifecycle management. WDYT?

File: pkg/controller/volume/persistentvolume/metrics/metrics.go [hunk 4]
```diff
@@ -198,14 +192,82 @@ func (collector *pvAndPVCCountCollector) pvcCollect(ch chan<- prometheus.Metric)
...
 		pluginName = "N/A"
 	}
-	if err != nil {
-		volumeOperationErrorsMetric.WithLabelValues(pluginName, opName).Inc()
+	volumeOperationErrorsMetric.WithLabelValues(pluginName, opName).Inc()
+}
+
+// operationTimestamp stores the start time of an operation by an operator
+type operationTimestamp struct {
+	operator  string
```
Comment (msau42) [inline pkg/controller/volume/persistentvolume/metrics/metrics.go]: Can we rename operator to pluginName to align with the actual metric field names?

File: pkg/controller/volume/persistentvolume/metrics/metrics.go [hunk 5]
```diff
@@ -198,14 +192,82 @@ func (collector *pvAndPVCCountCollector) pvcCollect(ch chan<- prometheus.Metric)
...
+	c.cache.LoadOrStore(key, ts)
+}
+
+// Delete deletes a value for a key.
+func (c *OperationStartTimeCache) Delete(key string) {
+	c.cache.Delete(key)
+}
+
+// Store creates an operationTimestamp with current time, and stores it for a key
+func (c *OperationStartTimeCache) Store(key, operatorName, operationName string) {
```
Comment (msau42) [inline pkg/controller/volume/persistentvolume/metrics/metrics.go]: is this used?

File: pkg/controller/volume/persistentvolume/pv_controller.go [hunk 6]
```diff
@@ -1309,11 +1345,30 @@ func (ctrl *PersistentVolumeController) provisionClaim(claim *v1.PersistentVolum
...
+		return nil
+	}
 	ctrl.scheduleOperation(opName, func() error {
-		pluginName, err := ctrl.provisionClaimOperation(claim)
-		timeTaken := time.Since(startTime).Seconds()
-		metrics.RecordVolumeOperationMetric(pluginName, "provision", timeTaken, err)
+		// create a start timestamp entry in cache for deletion operation if no one exists with
+		// key = claimKey, operator = provisionerName, operation = "provision"
+		claimKey := claimToClaimKey(claim)
+		ctrl.operationTimestamps.AddIfNotExist(claimKey, ctrl.getProvisionerName(plugin, storageClass), "provision")
```
Comment (msau42) [inline pkg/controller/volume/persistentvolume/pv_controller.go]: @davidz627 is the provision name going to be correct for the migration case?  we can address this separately if needed

File: pkg/controller/volume/persistentvolume/pv_controller.go [hunk 7]
```diff
@@ -1309,11 +1345,30 @@ func (ctrl *PersistentVolumeController) provisionClaim(claim *v1.PersistentVolum
...
+		// failed to find the requested provisioning plugin, directly return err for now.
+		// controller will retry the provisioning in every syncUnboundClaim() call
+		// retain the original behavior of returning nil from provisionClaim call
+		return nil
+	}
 	ctrl.scheduleOperation(opName, func() error {
-		pluginName, err := ctrl.provisionClaimOperation(claim)
-		timeTaken := time.Since(startTime).Seconds()
-		metrics.RecordVolumeOperationMetric(pluginName, "provision", timeTaken, err)
+		// create a start timestamp entry in cache for deletion operation if no one exists with
```
Comment (msau42) [inline pkg/controller/volume/persistentvolume/pv_controller.go]: nit: fix comment for provision operation

**Review by msau42 — COMMENTED** [review]
just some minor comments. otherwise lgtm

/approve

**Comment by msau42** [comment]
Will let @gnufied do final review

File: pkg/controller/volume/persistentvolume/pv_controller.go (hunk 7 above)
Comment (yuxiangqian) [reply pkg/controller/volume/persistentvolume/pv_controller.go]: done

File: pkg/controller/volume/persistentvolume/metrics/metrics.go (hunk 5 above)
Comment (yuxiangqian) [reply pkg/controller/volume/persistentvolume/metrics/metrics.go]: removed. thnx

File: pkg/controller/volume/persistentvolume/metrics/metrics.go (hunk 4 above)
Comment (yuxiangqian) [reply pkg/controller/volume/persistentvolume/metrics/metrics.go]: done

File: pkg/controller/volume/persistentvolume/pv_controller.go (hunk 1 above)
Comment (yuxiangqian) [reply pkg/controller/volume/persistentvolume/pv_controller.go]: updated "RecordMetric" as described above. when an operation has been successfully executed, i.e., err == nil, the corresponding cache'd entry will be removed within the function.

**Comment by yuxiangqian** [comment]
@gnufied ping :)

**Comment by gnufied** [comment]
Can you confirm if we have verified if the metrics are captured as expected after this change? There are existing e2es that capture these metrics. Do those needs to be updated - https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/volume_metrics.go#L136 ?

**Comment by yuxiangqian** [comment]
> Can you confirm if we have verified if the metrics are captured as expected after this change? There are existing e2es that capture these metrics. Do those needs to be updated - https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/volume_metrics.go#L136 ?

I have NOT run e2e tests by myself using kubetest, however the PR does not introduce any new metrics, and e2e tests at: https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/volume_metrics.go#L249 and https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/volume_metrics.go#L136  should have caught the issue. The PR does not change the existing behavior for error count, only the value of end to end latency might be changed for in-tree plugin when there's an temp error, and will change for external provider.

File: pkg/controller/volume/persistentvolume/pv_controller.go (hunk 6 above)
Comment (davidz627) [reply pkg/controller/volume/persistentvolume/pv_controller.go]: @ddebroy could you take a look

**Comment by yuxiangqian** [comment]
e2e testing with on GCP for csi driver results:
volume_operation_total_seconds_bucket{operation_name="delete",volume_plugin="csi-hostpath",le="0.1"} 4
volume_operation_total_seconds_bucket{operation_name="delete",volume_plugin="csi-hostpath",le="0.25"} 4
volume_operation_total_seconds_bucket{operation_name="delete",volume_plugin="csi-hostpath",le="0.5"} 4
volume_operation_total_seconds_bucket{operation_name="delete",volume_plugin="csi-hostpath",le="1"} 4
volume_operation_total_seconds_bucket{operation_name="delete",volume_plugin="csi-hostpath",le="2.5"} 4
volume_operation_total_seconds_bucket{operation_name="delete",volume_plugin="csi-hostpath",le="5"} 4
volume_operation_total_seconds_bucket{operation_name="delete",volume_plugin="csi-hostpath",le="10"} 4
volume_operation_total_seconds_bucket{operation_name="delete",volume_plugin="csi-hostpath",le="15"} 4
volume_operation_total_seconds_bucket{operation_name="delete",volume_plugin="csi-hostpath",le="25"} 4
volume_operation_total_seconds_bucket{operation_name="delete",volume_plugin="csi-hostpath",le="50"} 4
volume_operation_total_seconds_bucket{operation_name="delete",volume_plugin="csi-hostpath",le="120"} 4
volume_operation_total_seconds_bucket{operation_name="delete",volume_plugin="csi-hostpath",le="300"} 4
volume_operation_total_seconds_bucket{operation_name="delete",volume_plugin="csi-hostpath",le="600"} 4
volume_operation_total_seconds_bucket{operation_name="delete",volume_plugin="csi-hostpath",le="+Inf"} 4
volume_operation_total_seconds_sum{operation_name="delete",volume_plugin="csi-hostpath"} 0.161563064
volume_operation_total_seconds_count{operation_name="delete",volume_plugin="csi-hostpath"} 4
volume_operation_total_seconds_bucket{operation_name="provision",volume_plugin="csi-hostpath",le="0.1"} 4
volume_operation_total_seconds_bucket{operation_name="provision",volume_plugin="csi-hostpath",le="0.25"} 7
volume_operation_total_seconds_bucket{operation_name="provision",volume_plugin="csi-hostpath",le="0.5"} 7
volume_operation_total_seconds_bucket{operation_name="provision",volume_plugin="csi-hostpath",le="1"} 7
volume_operation_total_seconds_bucket{operation_name="provision",volume_plugin="csi-hostpath",le="2.5"} 7
volume_operation_total_seconds_bucket{operation_name="provision",volume_plugin="csi-hostpath",le="5"} 7
volume_operation_total_seconds_bucket{operation_name="provision",volume_plugin="csi-hostpath",le="10"} 7
volume_operation_total_seconds_bucket{operation_name="provision",volume_plugin="csi-hostpath",le="15"} 7
volume_operation_total_seconds_bucket{operation_name="provision",volume_plugin="csi-hostpath",le="25"} 7
volume_operation_total_seconds_bucket{operation_name="provision",volume_plugin="csi-hostpath",le="50"} 7
volume_operation_total_seconds_bucket{operation_name="provision",volume_plugin="csi-hostpath",le="120"} 7
volume_operation_total_seconds_bucket{operation_name="provision",volume_plugin="csi-hostpath",le="300"} 7
volume_operation_total_seconds_bucket{operation_name="provision",volume_plugin="csi-hostpath",le="600"} 7
volume_operation_total_seconds_bucket{operation_name="provision",volume_plugin="csi-hostpath",le="+Inf"} 7

for in-tree pd plug-in:
storage_operation_duration_seconds_bucket{operation_name="volume_provision",volume_plugin="kubernetes.io/gce-pd",le="0.1"} 0
storage_operation_duration_seconds_bucket{operation_name="volume_provision",volume_plugin="kubernetes.io/gce-pd",le="0.25"} 0
storage_operation_duration_seconds_bucket{operation_name="volume_provision",volume_plugin="kubernetes.io/gce-pd",le="0.5"} 0
storage_operation_duration_seconds_bucket{operation_name="volume_provision",volume_plugin="kubernetes.io/gce-pd",le="1"} 0
storage_operation_duration_seconds_bucket{operation_name="volume_provision",volume_plugin="kubernetes.io/gce-pd",le="2.5"} 1
storage_operation_duration_seconds_bucket{operation_name="volume_provision",volume_plugin="kubernetes.io/gce-pd",le="5"} 1
storage_operation_duration_seconds_bucket{operation_name="volume_provision",volume_plugin="kubernetes.io/gce-pd",le="10"} 1
storage_operation_duration_seconds_bucket{operation_name="volume_provision",volume_plugin="kubernetes.io/gce-pd",le="15"} 1
storage_operation_duration_seconds_bucket{operation_name="volume_provision",volume_plugin="kubernetes.io/gce-pd",le="25"} 1
storage_operation_duration_seconds_bucket{operation_name="volume_provision",volume_plugin="kubernetes.io/gce-pd",le="50"} 1
storage_operation_duration_seconds_bucket{operation_name="volume_provision",volume_plugin="kubernetes.io/gce-pd",le="120"} 1
storage_operation_duration_seconds_bucket{operation_name="volume_provision",volume_plugin="kubernetes.io/gce-pd",le="300"} 1
storage_operation_duration_seconds_bucket{operation_name="volume_provision",volume_plugin="kubernetes.io/gce-pd",le="600"} 1
storage_operation_duration_seconds_bucket{operation_name="volume_provision",volume_plugin="kubernetes.io/gce-pd",le="+Inf"} 1
storage_operation_duration_seconds_sum{operation_name="volume_provision",volume_plugin="kubernetes.io/gce-pd"} 2.109063131

File: pkg/controller/volume/persistentvolume/pv_controller.go (hunk 6 above)
Comment (gnufied) [reply pkg/controller/volume/persistentvolume/pv_controller.go]: provisioner name looks correct if migration is enabled. I do not see a problem here.

File: pkg/controller/volume/persistentvolume/pv_controller.go [hunk 8]
```diff
@@ -1011,11 +1041,17 @@ func (ctrl *PersistentVolumeController) reclaimVolume(volume *v1.PersistentVolum
 	case v1.PersistentVolumeReclaimDelete:
 		klog.V(4).Infof("reclaimVolume[%s]: policy is Delete", volume.Name)
 		opName := fmt.Sprintf("delete-%s[%s]", volume.Name, string(volume.UID))
-		startTime := time.Now()
+		// create a start timestamp entry in cache for deletion operation if no one exists with
+		// key = volume.Name, pluginName = provisionerName, operation = "delete"
+		ctrl.operationTimestamps.AddIfNotExist(volume.Name, ctrl.getProvisionerNameFromVolume(volume), "delete")
```
Comment (gnufied) [inline pkg/controller/volume/persistentvolume/pv_controller.go]: We aren't handling translated names correctly for delete operation - are we? it seems that, we should apply same logic we apply for provisioning operation. cc @davidz627

Comment (yuxiangqian) [reply pkg/controller/volume/persistentvolume/pv_controller.go]: the provisioner name for deletion is currently obtained from storage class's "provisioner" field. For provisioning, from my observation via e2e test, the same field is used for as the annotation "storage-provisioner" value for a PVC, as well as the annotation "provisioned-by" value of a PV.

Comment (davidz627) [reply pkg/controller/volume/persistentvolume/pv_controller.go]: Looks like `ctrl.getProvisionerName` does a translation if the volume is migrated, but `getProvisionerNameFromVolume` does not, they both should. Alternatively, could we just use: the `"pv.kubernetes.io/provisioned-by"` annotation which will be set correctly by migration path, instead of `GET`-ing the storage class inside these helper functions.

File: pkg/controller/volume/persistentvolume/pv_controller.go (hunk 6 above)
Comment (davidz627) [reply pkg/controller/volume/persistentvolume/pv_controller.go]: looks like `getProvisionerName` currently accounts for migration; however, `getProvisionerNameFromVolume` does not.

File: pkg/controller/volume/persistentvolume/pv_controller.go [hunk 9]
```diff
@@ -1661,3 +1726,29 @@ func (ctrl *PersistentVolumeController) findDeletablePlugin(volume *v1.Persisten
...
 }
+
+// obtain provisioner/deleter name for a volume
+func (ctrl *PersistentVolumeController) getProvisionerNameFromVolume(volume *v1.PersistentVolume) string {
+	storageClass := v1helper.GetPersistentVolumeClass(volume)
+	class, err := ctrl.classLister.Get(storageClass)
+	if err != nil {
+		return "N/A"
+	}
+	return class.Provisioner
```
Comment (davidz627) [inline pkg/controller/volume/persistentvolume/pv_controller.go]: this needs to have migration specific logic to return the correct provisioner name. Is there any reason why we're not using "pv.kubernetes.io/provisioned-by" annotation on the PV?

**Review by davidz627 — COMMENTED** [review]
only reviewed migration related comments

File: pkg/controller/volume/persistentvolume/pv_controller.go (hunk 9 above)
Comment (yuxiangqian) [reply pkg/controller/volume/persistentvolume/pv_controller.go]: thanks for clarifying. there was concern about "provisioned-by" could not be reliably used though my test seems showing that annotation is actually getting updated correctly by external provisioners. I am happy to switch to use "provisioned-by" annotation if its reliable. Does it also cover the in-tree case?

File: pkg/controller/volume/persistentvolume/pv_controller.go (hunk 8 above)
Comment (yuxiangqian) [reply pkg/controller/volume/persistentvolume/pv_controller.go]: @davidz627 for PV, is annotation "volume.beta.kubernetes.io/storage-provisioner" reliable enough as the annotation "pv.kubernetes.io/provisioned-by" for PV?

File: pkg/controller/volume/persistentvolume/pv_controller.go (hunk 9 above)
Comment (davidz627) [reply pkg/controller/volume/persistentvolume/pv_controller.go]: I think the old code used to use `provisioned-by` implicitly through the return value of `findDeletablePlugin`. I can't comment on its reliability for general external provisioners. You're right that maybe it's safer to use the storage class provisioner name, lets just do a translation of that then

Comment (yuxiangqian) [reply pkg/controller/volume/persistentvolume/pv_controller.go]: @davidz627 yes indeed as you commented, the original logic to look for plugin/provisioner for deletion uses the annotation ([check here](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/persistentvolume/pv_controller.go#L1636)) ONLY if it exists. otherwise, it uses volume spec for finding. That was the major reason why the reliability concern was raised above. I will keep the implementation as its for now then. thnx

**Comment by davidz627** [comment]
/hold
`getProvisionerNameFromVolume` still doesn't account for migration

**Comment by yuxiangqian** [comment]
> /hold
> `getProvisionerNameFromVolume` still doesn't account for migration

Is there anything else from this PR to accommodate for migration? @davidz627

File: pkg/controller/volume/persistentvolume/pv_controller.go (hunk 9 above)
Comment (davidz627) [reply pkg/controller/volume/persistentvolume/pv_controller.go]: I still think the storage class gotten here is going to be the in-tree one when migration is on. But I see we also dont have the plugin information to check whether this volume has been migrated. Could we use `provisioned-by` and fall back to the storage class provisioner if we can't find the `provisioned-by` annotation. (at least for migration cases this annotation should exist)

**Comment by davidz627** [comment]
@yuxiangqian 
Ah sorry I should clarify: https://github.com/kubernetes/kubernetes/pull/78061/files#r288724352

File: pkg/controller/volume/persistentvolume/pv_controller.go (hunk 9 above)
Comment (gnufied) [reply pkg/controller/volume/persistentvolume/pv_controller.go]: I think this has to be fixed correctly. i.e - we should fetch the `DeletablePlugin` from Provisioner name retrieved from storageclass. Check if migration is enabled for the plugin and use CSI name otherwise use provisioner name present in SC.

Comment (gnufied) [reply pkg/controller/volume/persistentvolume/pv_controller.go]: @msau42 also clarified that using `provisioned-by` could cause explosion of metric names for local-storage volumes because `provisioned-by` annotation points to node name in that case.

Comment (yuxiangqian) [reply pkg/controller/volume/persistentvolume/pv_controller.go]: @gnufied @davidz627 
instead of using storageclass defined provisioner name, to find the DeletablePlugin (if exists), I am proposing the following for getting the deleter name:
if has dynamic provisioned annotation, check the value of it, for external provisioner, use the value directly as the name, otherwise, use the plugin name to find DeletablePlugin and get the name.
If not dynamically provisioned, use spec to find plugin, falls back to "N/A" if none of the above.
WDYT?

Comment (yuxiangqian) [reply pkg/controller/volume/persistentvolume/pv_controller.go]: > @msau42 also clarified that using `provisioned-by` could cause explosion of metric names for local-storage volumes because `provisioned-by` annotation points to node name in that case.

"provision-by" is not used in this case for the concerns listed above

**Comment by yuxiangqian** [comment]
@davidz627 could you help reviewing the new logic to get provisioner name from a volume at [here?](https://github.com/kubernetes/kubernetes/pull/78061/files#diff-3f32ad90966ceeb297f57c2a8d726348R1731)
thanks

File: pkg/controller/volume/persistentvolume/pv_controller.go [hunk 10]
```diff
@@ -1661,3 +1726,47 @@ func (ctrl *PersistentVolumeController) findDeletablePlugin(volume *v1.Persisten
...
+func (ctrl *PersistentVolumeController) getProvisionerNameFromVolume(volume *v1.PersistentVolume) string {
+	plugin, err := ctrl.findDeletablePlugin(volume)
+	if err != nil {
+		return "N/A"
+	}
+	// if external provisioner was used for provisioning,
+	// the volume MUST have annotation of AnnDynamicallyProvisioned, use the value
+	// as the provisioner name
+	if plugin == nil {
+		return volume.Annotations[pvutil.AnnDynamicallyProvisioned]
```
Comment (davidz627) [inline pkg/controller/volume/persistentvolume/pv_controller.go]: I think in this case we can actually get it from the storage class no problem

File: pkg/controller/volume/persistentvolume/pv_controller.go [hunk 11]
```diff
@@ -1661,3 +1726,47 @@ func (ctrl *PersistentVolumeController) findDeletablePlugin(volume *v1.Persisten
...
+		if err != nil {
+			return "N/A"
+		}
+		provisionerName, err := ctrl.getCSINameFromIntreeName(class.Provisioner)
+		if err != nil {
+			return "N/A"
+		}
+		return provisionerName
+	}
+	return plugin.GetPluginName()
```
Comment (davidz627) [inline pkg/controller/volume/persistentvolume/pv_controller.go]: is `PluginName == ProvisionerName`? I think this is true for in-tree plugins but just want to double check

File: pkg/controller/volume/persistentvolume/pv_controller.go [hunk 12]
```diff
@@ -1661,3 +1726,47 @@ func (ctrl *PersistentVolumeController) findDeletablePlugin(volume *v1.Persisten
...
+	plugin, err := ctrl.findDeletablePlugin(volume)
+	if err != nil {
+		return "N/A"
+	}
+	// if external provisioner was used for provisioning,
+	// the volume MUST have annotation of AnnDynamicallyProvisioned, use the value
+	// as the provisioner name
+	if plugin == nil {
+		return volume.Annotations[pvutil.AnnDynamicallyProvisioned]
+	} else if plugin.IsMigratedToCSI() {
```
Comment (davidz627) [inline pkg/controller/volume/persistentvolume/pv_controller.go]: this case LGTM

File: pkg/controller/volume/persistentvolume/pv_controller.go (hunk 10 above)
Comment (yuxiangqian) [reply pkg/controller/volume/persistentvolume/pv_controller.go]: is it different from using this annotation? my understanding is its equivalent. Getting from storage class of course is another option, however I am trying to avoid the call to v1helper.GetPersistentVolumeClass if not necessary.

File: pkg/controller/volume/persistentvolume/pv_controller.go (hunk 11 above)
Comment (yuxiangqian) [reply pkg/controller/volume/persistentvolume/pv_controller.go]: for the last in-tree case, yes its equivalent I think

File: pkg/controller/volume/persistentvolume/pv_controller.go (hunk 10 above)
Comment (gnufied) [reply pkg/controller/volume/persistentvolume/pv_controller.go]: Using `AnnDynamicallyProvisioned` is still problematic though. As I said commented before - using `AnnDynamicallyProvisioned` here will result in arbitrary provisioner name(which is just node name in case of local-storage-provisioner) and will cause explosion of metrics for anyone using local-storage-provisioner. 

@yuxiangqian you commented:

>"provision-by" is not used in this case for the concerns listed above

Which I do not understand. Local storage plugin (in-tree) does not implement deletable interface and hence `plugin` will be `nil`. So either for local-storage use cases, depending on provisioner name specified in storageclass -either the metrics will be reported with `N/A` name or will cause explosion of metrics with each metric having node name as a dimension.

Comment (gnufied) [reply pkg/controller/volume/persistentvolume/pv_controller.go]: Since local-storage-provisioner tags PVs with `pv.kubernetes.io/provisioned-by=local-volume-provisioner-minikube-18f57fb2-a186-11e7-b543-080027d51893`, this will invariably result in each delete metric being reported separately..

Comment (yuxiangqian) [reply pkg/controller/volume/persistentvolume/pv_controller.go]: Good catch. By saying "provision-by is not used in this case for the concerns listed above" applies to the PVC thread not here. 
I will update the PR for plugin == nil case to use the storage class provisioner instead

File: pkg/controller/volume/persistentvolume/pv_controller.go [hunk 13]
```diff
@@ -1661,3 +1726,50 @@ func (ctrl *PersistentVolumeController) findDeletablePlugin(volume *v1.Persisten
...
+			// use the CSI name instead of in-tree name
+			provisionerName, err = ctrl.getCSINameFromIntreeName(class.Provisioner)
+			if err != nil {
+				return "N/A"
+			}
+		}
+		return provisionerName
+	}
+	return plugin.GetPluginName()
+}
```
Comment (gnufied) [inline pkg/controller/volume/persistentvolume/pv_controller.go]: Can this function be simplified to:

```golang
func (ctrl *PersistentVolumeController) getProvisionerNameFromVolume(volume *v1.PersistentVolume) string {
	plugin, _ := ctrl.findDeletablePlugin(volume)
	if plugin == nil {
		storageClass := v1helper.GetPersistentVolumeClass(volume)
		class, err := ctrl.classLister.Get(storageClass)
		if err != nil {
			return "N/A"
		}
		return class.Provisioner
	}
	if plugin.IsMigratedToCSI() {
		// use the CSI name instead of in-tree name
		provisionerName, err = ctrl.getCSINameFromIntreeName(class.Provisioner)
		if err != nil {
			return "N/A"
		}
		return provisionerName
	}
	return plugin.GetPluginName()
}
```

Comment (gnufied) [reply pkg/controller/volume/persistentvolume/pv_controller.go]: This is a minor nit though.. I don't mind either version of the function.

File: pkg/controller/volume/persistentvolume/pv_controller.go [hunk 14]
```diff
@@ -1661,3 +1726,50 @@ func (ctrl *PersistentVolumeController) findDeletablePlugin(volume *v1.Persisten
...
+		class, err := ctrl.classLister.Get(storageClass)
+		if err != nil {
+			return "N/A"
+		}
+		provisionerName := class.Provisioner
+		if plugin != nil {
+			// use the CSI name instead of in-tree name
+			provisionerName, err = ctrl.getCSINameFromIntreeName(class.Provisioner)
+			if err != nil {
+				return "N/A"
```
Comment (gnufied) [inline pkg/controller/volume/persistentvolume/pv_controller.go]: I know unlikely to happen but if we can't get CSI name for in-tree name, we should return `class.Provisioner` from here rather than `N/A`

File: pkg/controller/volume/persistentvolume/pv_controller.go (hunk 13 above)
Comment (yuxiangqian) [reply pkg/controller/volume/persistentvolume/pv_controller.go]: updated to return provisioner when getCSINameFromIntreeName fails

**Comment by gnufied** [comment]
@saad-ali @childsb can we add this PR to 1.15 milestone please?

**Comment by davidz627** [comment]
I'm concerned with the flakyness of these tests. @yuxiangqian have you confirmed the flakyness is due to issues outside of the scope of this PR?

**Comment by yuxiangqian** [comment]
> I'm concerned with the flakyness of these tests. @yuxiangqian have you confirmed the flakyness is due to issues outside of the scope of this PR?

I think so, I checked most test failures, i.e., for the GPU one it says some error like:
ERROR: (gcloud.compute.instances.create) Could not fetch resource:
ERROR: (gcloud.compute.instance-groups.managed.wait-until-stable) Timeout while waiting for group to become stable.

for the integration one:
I0530 19:30:31.814] dangling: []v1.OwnerReference{v1.OwnerReference{APIVersion:"mygroup.example.com/v1beta1", Kind:"foozpxh8a", Name:"ownervtkr7", UID:"b20b3559-cfeb-4a02-9c01-58bdc5736d21", Controller:(*bool)(nil), BlockOwnerDeletion:(*bool)(nil)}}
I0530 19:30:31.814] FAIL	k8s.io/kubernetes/test/integration/garbagecollector	558.064s

and see issues here:
https://github.com/kubernetes/kubernetes/issues/70180

At least I could not find anything which hints this PR caused issues like those. I have also conducted my own e2e test using gcp, none of those appears.

**Comment by yuxiangqian** [comment]
briefly checked some other PRs, like this one https://github.com/kubernetes/kubernetes/pull/78541, it seems blazel-test and integration tests fails from time to time.

**Comment by davidz627** [comment]
sgtm thanks for the detailed answer @yuxiangqian 
/retest

File: pkg/controller/volume/persistentvolume/pv_controller.go [hunk 15]
```diff
@@ -1661,3 +1726,48 @@ func (ctrl *PersistentVolumeController) findDeletablePlugin(volume *v1.Persisten
 	}
 	return plugin, nil
 }
+
+// obtain provisioner/deleter name for a volume
+func (ctrl *PersistentVolumeController) getProvisionerNameFromVolume(volume *v1.PersistentVolume) string {
+	plugin, err := ctrl.findDeletablePlugin(volume)
+	if err != nil {
+		return "N/A"
```
Comment (msau42) [inline pkg/controller/volume/persistentvolume/pv_controller.go]: Local PV will end up with "N/A" as the plugin since it's not dynamically provisioned and the plugin itself doesn't implement delete.

Should we let it fall through instead of returning "N/A" here?
